libindy/src/services/pool/catchup.rs

```diff
@@ -115,40 +118,67 @@ impl CatchupHandler {
         if self.pending_catchup.is_some() {
             return Ok(CatchupProgress::InProgress);
         }
-        let mut votes: HashMap<(String, usize), usize> = HashMap::new();
-        for node_vote in &self.nodes_votes {
-            if let &Some(ref node_vote) = node_vote {
-                let cnt = *votes.get(&node_vote).unwrap_or(&0) + 1;
-                votes.insert((node_vote.0.clone(), node_vote.1), cnt);
-            }
-        }
+
+        let votes: HashMap<(String, usize, Option<Vec<String>>), usize> = self.nodes_votes.iter().cloned()
+            .filter_map(|e| e)
+            .fold(HashMap::new(), |mut acc, vote| {
+                let cnt = acc.get(&vote).unwrap_or(&0) + 1;
+                acc.insert(vote, cnt);
+                acc
+            });
+
         if let Some((most_popular_vote, votes_cnt)) = votes.iter().max_by_key(|entry| entry.1) {
             if *votes_cnt == self.nodes.len() - self.f {
-                let &(ref target_mt_root, target_mt_size) = most_popular_vote;
-                let cur_mt_size = self.merkle_tree.count();
-                let cur_mt_hash = self.merkle_tree.root_hash().to_base58();
-                if target_mt_size == cur_mt_size {
-                    if cur_mt_hash.eq(target_mt_root) {
-                        return Ok(CatchupProgress::NotNeeded);
-                    } else {
-                        return Err(PoolError::CommonError(CommonError::InvalidState(
-                            "Ledger merkle tree doesn't acceptable for current tree.".to_string())));
+                return match self._try_to_catch_up(most_popular_vote) {
+                    //WARNING: If we receive invalid cache then it means that cache was cleared. We need to rebuild merkle tree and try to catch up once again
```

Comment:
I suggest to try to pop cache cleaning to this method. like
```
self._try_to_catch_up().or_else(|err| {
  drop_cache() // here or inside next if
  if err == InvalidCache { try_again() } else { err } 
})
```
It may be more clear and don't require WARNING comment